File: backend/src/main/java/com/hospital/backend/service/NotificationService.java
package com.hospital.backend.service;

import com.hospital.backend.dto.NotificationResponse;
import com.hospital.backend.model.Notification;

import java.util.List;

public interface NotificationService {

    // Create and send notification
    Notification createNotification(Notification notification);

    // Send notification to user via WebSocket
    void sendNotification(String userId, Notification notification);

    // Get all notifications for user
    List<NotificationResponse> getNotifications(String userId);

    // Get unread notifications for user
    List<NotificationResponse> getUnreadNotifications(String userId);

    // Get unread count
    long getUnreadCount(String userId);

    // Mark notification as read
    void markAsRead(String notificationId);

    // Mark all as read
    void markAllAsRead(String userId);

    // Delete notification
    void deleteNotification(String notificationId);

    // Delete all notifications for user
    void deleteAllNotifications(String userId);

    // Notify about appointment
    void notifyAppointmentCreated(String appointmentId, String patientId, String doctorId, String appointmentTime);

    void notifyAppointmentUpdated(String appointmentId, String patientId, String doctorId);

    void notifyAppointmentCancelled(String appointmentId, String patientId, String doctorId);

    // Notify about prescription
    void notifyPrescriptionCreated(String prescriptionId, String patientId, String doctorId);

    // Notify about reports
    void notifyReportAvailable(String reportId, String patientId);

    // System notification
    void sendSystemNotification(String userId, String title, String message);
}

File: frontend/services/notificationService.js
"use client";

import { API_BASE } from "../utils/constants";

/**
 * Get all notifications
 */
export async function getNotifications(token) {
  const res = await fetch(`${API_BASE}/notifications`, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
    },
  });

  const data = await res.json();

  if (!res.ok) {
    throw new Error(data?.message || data?.error || "Failed to fetch notifications");
  }

  return data;
}

/**
 * Get unread notifications only
 */
export async function getUnreadNotifications(token) {
  const res = await fetch(`${API_BASE}/notifications/unread`, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
    },
  });

  const data = await res.json();

  if (!res.ok) {
    throw new Error(data?.message || data?.error || "Failed to fetch unread notifications");
  }

  return data;
}

/**
 * Get unread notification count
 */
export async function getUnreadCount(token) {
  const res = await fetch(`${API_BASE}/notifications/unread/count`, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
    },
  });

  const data = await res.json();

  if (!res.ok) {
    throw new Error(data?.message || data?.error || "Failed to fetch unread count");
  }

  return data.unreadCount;
}

/**
 * Mark notification as read
 */
export async function markAsRead(notificationId, token) {
  const res = await fetch(`${API_BASE}/notifications/${notificationId}/read`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
    },
  });

  const data = await res.json();

  if (!res.ok) {
    throw new Error(data?.message || data?.error || "Failed to mark notification as read");
  }

  return data;
}

/**
 * Mark all notifications as read
 */
export async function markAllAsRead(token) {
  const res = await fetch(`${API_BASE}/notifications/read-all`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
    },
  });

  const data = await res.json();

  if (!res.ok) {
    throw new Error(data?.message || data?.error || "Failed to mark all as read");
  }

  return data;
}

/**
 * Delete notification
 */
export async function deleteNotification(notificationId, token) {
  const res = await fetch(`${API_BASE}/notifications/${notificationId}`, {
    method: "DELETE",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
    },
  });

  const data = await res.json();

  if (!res.ok) {
    throw new Error(data?.message || data?.error || "Failed to delete notification");
  }

  return data;
}

/**
 * Delete all notifications
 */
export async function deleteAllNotifications(token) {
  const res = await fetch(`${API_BASE}/notifications`, {
    method: "DELETE",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
    },
  });

  const data = await res.json();

  if (!res.ok) {
    throw new Error(data?.message || data?.error || "Failed to delete all notifications");
  }

  return data;
}
